pkg/operator/ceph/object/notification/controller.go

```diff
@@ -0,0 +1,227 @@
+	if !cephBN.GetDeletionTimestamp().IsZero() {
+		logger.Debugf("CephBucketNotification: %q was deleted", cephBN.Name)
+
+		// Return and do not requeue. Successful deletion.
+		return reconcile.Result{}, nil
+	}
+
+	// Populate clusterInfo during each reconcile
+	r.clusterInfo, _, _, err = mon.LoadClusterInfo(r.context, cephCluster.Namespace)
+	if err != nil {
+		return reconcile.Result{}, errors.Wrap(err, "failed to populate cluster info")
+	}
+
+	// create the notification if needed
+	if err = r.updateCephBucketNotifications(cephBN); err != nil {
+		return reconcile.Result{}, errors.Wrapf(err, "failed to create bucket notifications")
+	}
+	// Return and do not requeue
+	logger.Debug("CephBucketNotification done reconciling")
+	return reconcile.Result{}, nil
+}
+
+func (r *ReconcileNotifications) updateCephBucketNotifications(cephBN *cephv1.CephBucketNotification) error {
+	// step 1: fetch all OBCs that has a label matching this CBN
+	namespace := cephBN.Namespace
+	namespaceListOpt := client.InNamespace(namespace)
+	labelListOpt := client.MatchingLabels{
+		notificationLabelPrefix + cephBN.Name: cephBN.Name,
+	}
+	cephOBCList := &bktv1alpha1.ObjectBucketClaimList{}
+	err := r.client.List(context.TODO(), cephOBCList, namespaceListOpt, labelListOpt)
+	if err != nil {
+		return err
+	}
+	if len(cephOBCList.Items) == 0 {
+		logger.Debugf("no ObjectbucketClaim associated with notification: %q", cephBN.Name)
+		return nil
+	}
+
+	// step 2: get the topic associated with the notification
+	bucketTopic := &cephv1.CephBucketTopic{}
+	if err := r.client.Get(context.TODO(), types.NamespacedName{Namespace: namespace, Name: cephBN.Spec.Topic}, bucketTopic); err != nil {
+		return errors.Wrapf(err, "failed to retrieve CephBucketTopic %q", cephBN.Spec.Topic)
+	}
+	topicARN, err := topic.GetARN(bucketTopic)
+	if err != nil {
+		return errors.Wrap(err, "failed to get topic ARN")
+	}
+
+	// step 3: loop through all OBCs in the list and get their OB and SC
+	for _, cephOBC := range cephOBCList.Items {
+		cephOB := bktv1alpha1.ObjectBucket{}
+		if err := r.client.Get(context.TODO(), types.NamespacedName{Namespace: namespace, Name: cephOBC.Spec.ObjectBucketName}, &cephOB); err != nil {
+			return errors.Wrapf(err, "failed to retrieve ObjectBucket %q", cephOBC.Spec.ObjectBucketName)
+		}
+		cephSC := storagev1.StorageClass{}
+		if err := r.client.Get(context.TODO(), types.NamespacedName{Namespace: namespace, Name: cephOBC.Spec.StorageClassName}, &cephSC); err != nil {
+			return errors.Wrapf(err, "failed to retrieve StorageClass %q", cephOBC.Spec.StorageClassName)
+		}
+
+		// step 4: provision the notification
+		p := Provisioner{
+			Client:      r.client,
+			Context:     r.context,
+			ClusterInfo: r.clusterInfo,
+			ClusterSpec: r.clusterSpec,
+		}
+		if err := p.CreateSession(
+			cephOB.Spec.AdditionalState["cephUser"],
+			types.NamespacedName{
+				Name:      cephSC.Parameters["objectStoreName"],
+				Namespace: cephSC.Parameters["objectStoreNamespace"],
```

Comment:
These should come from the CephOBN spec now right?

Reply:
i don't think we should add the storage identity to the CBN. 
I thought that the notification and the bucket must be in the same object store anyway?

Reply:
It's my understanding the CBN and OBC will be in the same namespace, yes. The problem is that the StorageClass can be modified or deleted after the OBC is provisioned, so we can't rely on the SC having accurate info. We will have to get this info from the OB/OBC or have it provided in the CBN.

Reply:
From the OB:

`bucket-provisioner=<namespace>.ceph.rook.io-bucket` identifies the namespace of the CephObjectStore. E.g., `bucket-provisioner=rook-ceph.ceph.rook.io-bucket` is in the namespace `rook-ceph`.

`rook-ceph-rgw-<store-name>.<namespace>.svc` actually identifies the name of the CephObjectStore and namespace. E.g., `rook-ceph-rgw-my-store.rook-ceph.svc` is store `my-store` in namespace `rook-ceph`.

Reply:
thanks!  will use this instead of SC
will leave external clusters to be out-of-scope for now